### PR TITLE
Added tracking for deleted namespace status check in restore flow 

### LINKED ISSUE #8234: Terminating namespace polling for each resource during restore
[EnsureNamespaceExistsAndIsReady](https://github.com/vmware-tanzu/velero/blob/main/pkg/util/kube/utils.go#L70)

Problem: During restore for every resource within the namespace, we are calling the check to await on if namespace exists (wait for 10 min polling). For instance, if we have 100 resources in a namespace that itself is in terminating state for very long then it impacts the restore flow to get stuck/halt and it too increases the time to restore as it does it for each resource in the same namespace.

_Originally posted by @mayankagg9722 in https://github.com/vmware-tanzu/velero/issues/7516#issuecomment-2022457069_

### PR DESCRIPTION
**Pull request by sangitaray2021**
Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #8234

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file (`make new-changelog`)](https://velero.io/docs/main/code-standards/#adding-a-changelog) or comment `/kind changelog-not-required` on this PR.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by anshulahuja98** [comment]
For - #8234

**Review by kaovilai — CHANGES_REQUESTED** [review]
Please add changelog and run linter.

**Comment by reasonerjt** [comment]
Given we've reached FC and this may impact the restore flow, let's hold this one until the branch is cut for v1.15

**Comment by anshulahuja98** [comment]
@kaovilai can you help review the PR?

**Review by kaovilai — COMMENTED** [review]
IIUC from reviewing, this change reduce potential duplicate `await on if namespace exists (wait for 10 min polling)` to just one instance per restore. Is that correct?

It's not meant to add polling until deletion is complete to resume some other process?

**Comment by sangitaray2021** [comment]
> IIUC from reviewing, this change reduce potential duplicate `await on if namespace exists (wait for 10 min polling)` to just one instance per restore. Is that correct?
> 
> It's not meant to add polling until deletion is complete to resume some other process?

yes thats correct. We are reducing duplicate wait by caching the namespace which is under deletion.

File: pkg/util/kube/namespace_deletionstatus_tracker.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+// NamespaceDeletionStatusTracker keeps track of in-progress backups.
+type NamespaceDeletionStatusTracker interface {
+	// Add informs the tracker that a polling is in progress to check namespace deletion status.
+	Add(ns, name string)
+	// Delete informs the tracker that a namespace deletion is completed.
+	Delete(ns, name string)
+	// Contains returns true if the tracker is tracking the namespace deletion progress.
+	Contains(ns, name string) bool
+}
```
Comment (kaovilai) [inline pkg/util/kube/namespace_deletionstatus_tracker.go:34]: Why are there both ns and name parameters?
What purpose does that serve other than duplicate value params?

Is this interface meant to be more generic? Perhaps `namespaceDeletionStatusTrackerKey(ns, name)` was meant to be more like `resourceDeletionStatusTrackerKey(resource, ns, name string) string` ?

File: pkg/util/kube/resource_deletionstatus_tracker.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+package kube
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// resourceDeletionStatusTracker keeps track of in-progress backups.
```
Comment (kaovilai) [inline pkg/util/kube/resource_deletionstatus_tracker.go:26]:
```suggestion
// resourceDeletionStatusTracker keeps track of items pending deletion blocking in-progress backups.
```

**Review by kaovilai — APPROVED** [review]
lgtm

File: pkg/util/kube/resource_deletionstatus_tracker.go (hunk 2 above)
Comment (anshulahuja98) [reply pkg/util/kube/resource_deletionstatus_tracker.go:26]: @sangitaray2021, please add this also.

File: pkg/util/kube/utils.go [hunk 3]
```diff
@@ -107,7 +114,12 @@ func EnsureNamespaceExistsAndIsReady(namespace *corev1api.Namespace, client core
 	// err will be set if we timed out or encountered issues retrieving the namespace,
 	if err != nil {
 		if terminatingNamespace {
+			// If the namespace is marked for deletion, and we timed out, adding it in tracker
+			resourceDeletionStatusTracker.Add(namespace.Kind, namespace.Name, namespace.Name)
 			return false, nsCreated, errors.Wrapf(err, "timed out waiting for terminating namespace %s to disappear before restoring", namespace.Name)
+		} else if namespaceAlreadyInDeletionTracker {
+			// If the namespace is already in the tracker, return an error.
+			return false, nsCreated, err
```
Comment (anshulahuja98) [inline pkg/util/kube/utils.go:122]: let's make this error more verbose. 
Something like - Skipping polling for terminating namespace.

**Review by anshulahuja98 — APPROVED** [review]

**Comment by reasonerjt** [comment]
@anshulahuja98 I don't quite understand the background of the problem.  I'm asking Xun @blackpiglet to take a look as he used to have discussion around this topic.

**Comment by blackpiglet** [comment]
Per my understanding, this PR tracks the namespace hanging in the terminating state.
If the namespace already hits a timeout during the restore, it is added to the tracked, and a flag is marked as true.
If there is a second time meeting the same timeout issue for the namespace, the restore is marked as failed.

IMO, revoking the ongoing restore is the ideal solution for this issue, but that obviously needs more effort, so I think this is also an acceptable solution for now.

**Comment by anshulahuja98** [comment]
@blackpiglet can you signoff if the PR looks good?

**Review by blackpiglet — APPROVED** [review]

**Pull request merged**